### ISSUE TITLE
Slow Dart Dev incremental compilation

### ISSUE THREAD

**Issue by mnordine** [dart-lang/sdk#50664]
Using `cloc` for getting lines of code for our private repo gives:
```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Dart                           433          19854           2477          73490
```
After running `$ dart pub global run webdev serve web:53323`:
```
Building package executable... (1.1s)
Built webdev:webdev.
[WARNING] Throwing away cached asset graph because the build phases have changed. This most commonly would happen as a result of adding a new dependency or updating your dependencies.
[WARNING] Throwing away cached asset graph because the language version of some package(s) changed. This would most commonly happen when updating dependencies or changing your min sdk constraint.
[INFO] Cleaning up outputs from previous builds. completed, took 286ms
[INFO] Building new asset graph completed, took 2.1s
[INFO] Checking for unexpected pre-existing outputs. completed, took 0ms
[INFO] Serving `web` on http://127.0.0.1:53323
[INFO] Generating SDK summary completed, took 2.0s
[INFO] Running build completed, took 25.8s
[INFO] Caching finalized dependency graph completed, took 426ms
[INFO] Succeeded after 26.2s with 5418 outputs (11599 actions)
```

Then, changing a single character in our program yields:
```
[INFO] Updating asset graph completed, took 12ms
[INFO] Running build completed, took 13.5s
[INFO] Caching finalized dependency graph completed, took 387ms
[INFO] Succeeded after 13.9s with 326 outputs (322 actions)
```

Note that this is using a 12-core, 24-thread AMD Ryzen 3900X. On lesser machines we sometimes get times in the 30-60 second range.

Dart 2.18.4

@vsmenon 

Update: Should mention, Windows 11.

**Comment by vsmenon** [comment]
@jakemac53 may have some tips on how to look at the asset graph.  And possibly correct some of the below.  :-)

Roughly speaking, webdev+DDC compiles at the granularity of strongly connected components (based on file imports).  webdev builds a graph of strongly connected components (each consisting of files that all cyclically import each other.  If you change a file, it invokes DDC on the entire component.  It also then transitively invokes DDC on components that dependent on the modified one.

So, changing a leaf file that itself is small and has no downstream dependencies should be quick.  Changing a file that's part of a big component and has downstream dependencies will be slower.

That's a roundabout way of saying that restructuring the package - if possible - might help a lot.

**Comment by jakemac53** [comment]
Exactly what @vsmenon said - this is likely based on either the import structure of the app causing it all to be compiled as one module, or the particular file you are choosing to edit, and the particular type of edit being made.

It could also be related to other codegen, if you have any.

You can get a rough idea of where time is going in your build by following these instructions https://github.com/dart-lang/build/blob/master/docs/measuring_performance.md.

**Comment by mnordine** [comment]
Reading that document, I have no idea what command to run with those flags. I've tried many combinations, but none work. I also don't see either of those flags in `--help`. In general, when someone tells me about a flag, it seems like it's not documented, and I have no idea where to put the flag. After `dart`? `pub`? `run`? `webdev`? `serve`? etc...

**Comment by jakemac53** [comment]
Ah, if you are using webdev it gets a bit weird, it may not support these flags. I would use build_runner directly, and the arguments should be documented in its CLI.

**Comment by jakemac53** [comment]
tldr; `dart run build_runner serve --track-performance` and then load `http://localhost:8080/$perf` (port may be different)

**Comment by vsmenon** [comment]
@kevmoo  @atsansone @MaryaBelanger - can we fix docs here?  

Do we want to recommend `dart run build_runner serve` over `webdev` in all cases now?

**Comment by jakemac53** [comment]
I can't remember for sure if webdev supports plumbing arbitrary flags through to build_runner, but if so we could just document that better, and maybe even make the `--help` flag also print the output of the build_runner `--help` flag for the same command?

**Comment by atsansone** [comment]
@vsmenon : `dart run build_runner serve` is a lot more pain than `webdev` to type. I don't know what advantage `drbrs` offers than `webdev`. I do know that in the [commit that made the last round of changes](https://github.com/dart-lang/site-www/pull/4203/commits/0cd9104d5cfaf853f99d22a6ff647349da95c02d), `webdev` was said to make the decision between development and production builds for you. If that's the case, then I would stick with `webdev`. This may be something I can discuss with a SWE to get more information.

**Comment by jakemac53** [comment]
Yeah webdev is a more opinionated tool (and more targeted at web development specifically, hence the dwds/devtools integration). Where build_runner is general purpose build system and makes no assumptions.

**Comment by jakemac53** [comment]
I don't know that we actually want to tell users to stop using `webdev`. We could certainly, but such a decision hasn't been made and it isn't obvious that would be the best choice.

`webdev` also provides a nice consistent UX even with underlying changes to the build system - for instance a switch to use frontend server for DDC compilation.

**Comment by jakemac53** [comment]
> What is that and how do I use it? Is it the long-running process I've seen in commits?

It isn't supported currently outside of flutter - but it is what flutter web uses in dev mode. There isn't currently any plan to migrate `webdev` over to that. It might be possible to use the flutter SDK to compile/serve a Dart web app but I haven't tried, and also it doesn't integrate very nicely with build_runner today.

**Comment by mnordine** [comment]
> It isn't supported currently outside of flutter

What's the rationale for only flutter?

**Comment by jakemac53** [comment]
> What's the rationale for only flutter?

The simple answer to that support was added by the flutter team, for flutter. It is a non-trivial amount of work to support it in webdev, and there are also a large number of tradeoffs. Dart web projects also tend to require codegen, where flutter does not, and the story there isn't nearly as great (there basically isn't one, it would have to be invented).

But, I do think this issue is getting derailed quite a bit at this point, feel free to file a separate issue to discuss using the frontend server for web compilation though? From my perspective this would best be integrated directly into the `dart` tool at this point anyways, and `webdev` would wrap that instead of build_runner. But that is a whole different rabbit hole.

**Comment by sigmundch** [comment]
After a quick chat with @jakemac53, I decided it makes sense to close this issue. I also filed https://github.com/dart-lang/webdev/issues/1827 to track the general feature request to consider using frontend-server for non-flutter use cases. Besides that, there is still the original question here to be addressed: to improve performance on your specific scenario. Hopefully the tips above to profile will be helpful. Feel free to reply again here if you find additional data or if there is anything else we can do to help.

**Comment by mnordine** [comment]
@sigmundch sgtm. I'm still not sure how to fix my issue with the info given, I have a very complex app. I'll just start experimenting and see where I get.

**Comment by sigmundch** [comment]
Thanks @mnordine, keep us posted!

Feel free to share any info you feel it's OK to post here from the profiles. Hopefully we can narrow down to the probable cause and potential solutions. For example:
* if a module is too big and expensive to rebuild, maybe it is possible to split it into smaller pieces
* if one kind of sourcegen action is especially slow, potentially there is follow up that can be done to improve that action, or isolate some kind of input that is making that action particularly slow.

**Comment by vsmenon** [comment]
@mnordine - you may want to try your app view the `flutter` - e.g., `flutter run -d chrome`.  There is no requirement that the Dart code imports or uses Flutter.  I sometimes use that to run simple Dart pure web apps.  As alluded to above, the flutter tooling is just more resourced and optimized at this point.

That said, I'm not sure if anyone has used it like that at scale.  It doesn't support all forms of codegen (e.g., AngularDart) but it does others (built_value, json_serializable).

**Comment by mnordine** [comment]
Thanks @vsmenon, will give that a try
